### ISSUE TITLE
https://firecow.github.io/gitlab-ci-local/ppa/pubkey.gpg is expired

### ISSUE THREAD

**Issue by malem-robotiq** [firecow/gitlab-ci-local#734]
sudo apt-get
Get:2 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease [2111 B]                                                                                                                                                                                                                                                                                                                                              
Err:2 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease                                                                                                                                                                              
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
Reading package lists... Done
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562

**Comment by malem-robotiq** [comment]
workaround:
sudoedit /etc/apt/sources.list.d/gitlab-ci-local.list
`deb [allow-insecure=yes] https://firecow.github.io/gitlab-ci-local/ppa/ ./`

**Comment by firecow** [comment]
Why on earth am I not seeing this issue.

Please disclose more information about your OS.

**Comment by malem-robotiq** [comment]
Hi, thanks for the quick answer!

And I forgot to thank you for your wonderful project!

OS is ubuntu 20.04

here is what I tried
`
user@...:~$ curl -s "https://firecow.github.io/gitlab-ci-local/ppa/pubkey.gpg" | sudo apt-key add -
OK  
user@...:~$ sudo wget -O /etc/apt/sources.list.d/gitlab-ci-local.sources https://firecow.github.io/gitlab-ci-local/ppa/gitlab-ci-local.sources
--2023-01-25 11:36:28--  https://firecow.github.io/gitlab-ci-local/ppa/gitlab-ci-local.sources
Resolving firecow.github.io (firecow.github.io)... 185.199.110.153, 185.199.108.153, 185.199.109.153, ...
Connecting to firecow.github.io (firecow.github.io)|185.199.110.153|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3383 (3.3K) [application/octet-stream]
Saving to: '/etc/apt/sources.list.d/gitlab-ci-local.sources'

/etc/apt/sources.list.d/gitlab-ci-loca 100%[============================================================================>]   3.30K  --.-KB/s    in 0s      

2023-01-25 11:36:28 (41.5 MB/s) - '/etc/apt/sources.list.d/gitlab-ci-local.sources' saved [3383/3383]

user@...:~$ sudo apt update
E: Invalid value set for option Signed-By regarding source https://firecow.github.io/gitlab-ci-local/ppa/ ./ (not a fingerprint)
E: The list of sources could not be read.
user@...:~$ sudo rm /etc/apt/sources.list.d/gitlab-ci-local.sources 
user@...:~$ sudo apt update               
[...]
Fetched 1242 kB in 2s (749 kB/s)    
Reading package lists... Done
Building dependency tree       
Reading state information... Done
28 packages can be upgraded. Run 'apt list --upgradable' to see them.
user@...:~$ sudo wget -O /etc/apt/sources.list.d/gitlab-ci-local.list https://firecow.github.io/gitlab-ci-local/ppa/gitlab-ci-local.list
--2023-01-25 11:39:16--  https://firecow.github.io/gitlab-ci-local/ppa/gitlab-ci-local.list
Resolving firecow.github.io (firecow.github.io)... 185.199.110.153, 185.199.108.153, 185.199.109.153, ...
Connecting to firecow.github.io (firecow.github.io)|185.199.110.153|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 104 [text/plain]
Saving to: '/etc/apt/sources.list.d/gitlab-ci-local.list'

/etc/apt/sources.list.d/gitlab-ci-loca 100%[============================================================================>]     104  --.-KB/s    in 0s      

2023-01-25 11:39:16 (12.4 MB/s) - '/etc/apt/sources.list.d/gitlab-ci-local.list' saved [104/104]

user@...:~$ sudo apt update
Get:2 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease [2111 B]                                                                                  
[...]                                                                                          
Err:2 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease                                                                                           
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
[...]
Reading package lists... Done
W: GPG error: https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
E: The repository 'https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
`

**Comment by malem-robotiq** [comment]
so I correct myself, I assumed the gpg key was expired. but it might be something else. This very method used to work a month ago.

**Comment by malem-robotiq** [comment]
ok I close the issue I was missing the `| gpg --dearmor`

sorry for the trouble

**Comment by malem-robotiq** [comment]
after all it is still not working
```
$ curl -s "https://firecow.github.io/gitlab-ci-local/ppa/pubkey.gpg" | gpg --dearmor | sudo apt-key add -
OK
$ sudo apt-key list | grep gitlab-ci-local
Warning: apt-key output should not be parsed (stdout is not a terminal)
uid           [ unknown] Mads Jon Nielsen (Key for destributing gitlab-ci-local binary) <madsjon@gmail.com>
$ sudo apt update
Hit:2 http://security.ubuntu.com/ubuntu focal-security InRelease                                                                                           
Get:3 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease [2111 B]                                                                                  
[...]                                                                                
Get:7 http://ca.archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]                                                                               
Hit:8 https://download.docker.com/linux/ubuntu focal InRelease                                                                                             
Hit:9 https://packages.microsoft.com/repos/vscode stable InRelease                                                                                         
Hit:10 http://archive.ubuntu.com/ubuntu focal InRelease                                                                                                    
Hit:1 https://apt.llvm.org/focal llvm-toolchain-focal-14 InRelease                                                                                         
Hit:11 http://lenovo.archive.canonical.com focal InRelease                                                                                                 
Err:3 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease                                                                                      
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
[...]
Reading package lists... Done
W: GPG error: https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
E: The repository 'https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.

-

**Comment by malem-robotiq** [comment]
from https://wiki.debian.org/DebianRepository/UseThirdParty

Troubleshooting

`W: GPG error: http://deriv.example.net stable InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 744ACF4DF3319FFA`

This error message is commonly seen when the OpenPGP certificate you are using is ASCII-Armored instead of being in binary form.

**Comment by malem-robotiq** [comment]
```
$ sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 6504D9F3E15AB562
Executing: /tmp/apt-key-gpghome.YqG4evCj7S/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 6504D9F3E15AB562
gpg: keyserver receive failed: No data
$ sudo apt-key adv --keyserver keyring.debian.org --recv-keys 6504D9F3E15AB562
Executing: /tmp/apt-key-gpghome.WVxK2X2KgQ/gpg.1.sh --keyserver keyring.debian.org --recv-keys 6504D9F3E15AB562
gpg: no valid OpenPGP data found.
gpg: Total number processed: 0
```

**Comment by malem-robotiq** [comment]
Ok I found the problem. it was caused by this commit https://github.com/firecow/gitlab-ci-local/commit/c92ed05d7f775281a6d9c2444e45f027f60be6a4

First, on my ubuntu 20.04, the usr/apt/keyrings folder does not exist, the standard place to put alien keyrings is /urs/share/keyrings I think.

Here is what I did:

```
$ curl https://firecow.github.io/gitlab-ci-local/ppa/pubkey.gpg | sudo gpg -o /usr/share/keyrings/gitlab-ci-local.gpg --dearmor
$ echo "deb [arch=amd64, signed-by=/usr/share/keyrings/gitlab-ci-local.gpg] https://firecow.github.io/gitlab-ci-local/ppa ./" | sudo tee /etc/apt/sources.list.d/gitlab-ci-local.list > /dev/null
```

**Comment by malem-robotiq** [comment]
in fact, the most future-proof way to fix this is to download the key to an asc in trusted.gpg.d and use this asc as signed-by arg

here would be a command line
```
$ curl https://firecow.github.io/gitlab-ci-local/ppa/pubkey.gpg | sudo gpg -o /usr/apt/trusted.gpg.d/gitlab-ci-local.asc
[... replace the following if you change your hosted .list...]
$ echo "deb [arch=amd64, signed-by=/usr/apt/trusted.gpg.d/gitlab-ci-local.gpg] https://firecow.github.io/gitlab-ci-local/ppa ./" | sudo tee /etc/apt/sources.list.d/gitlab-ci-local.list > /dev/null
```
here is my ansible job

```
- name: gitlab-ci-local | avoid apt_key once it is removed from your distro
  block:
    - name: gitlab-ci-local | no apt key
      get_url:
        url: https://firecow.github.io/gitlab-ci-local/ppa/pubkey.gpg
        dest: /etc/apt/trusted.gpg.d/gitlab-ci-local.asc

    - name: gitlab-ci-local | apt source
      apt_repository:
        repo: "deb [signed-by=/etc/apt/trusted.gpg.d/gitlab-ci-local.asc] https://firecow.github.io/gitlab-ci-local/ppa ./"
        filename: gitlab-ci-local.list
        state: present
  become: yes
```

**Comment by firecow** [comment]
@malem-robotiq Thanks for the debug.

I'll try out the new install instructions on a Ubuntu 20.04, and adjust the readme if it has any inconsistentcies.

**Comment by firecow** [comment]
```bash
Hit:18 https://pkg.cloudflareclient.com focal InRelease                                       
Err:13 https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
Hit:17 https://packagecloud.io/slacktechnologies/slack/debian jessie InRelease
Reading package lists... Done
W: GPG error: https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6504D9F3E15AB562
E: The repository 'https://firecow.github.io/gitlab-ci-local/ppa ./ InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```



A yes, I can reproduce, there's definently a bug in the readme instructions

**Comment by firecow** [comment]
Fixed in bd5e6d547363d9514034d71f1a728fcb1254e3c1

**Comment by malem-robotiq** [comment]
thank you!

**Comment by ecourreges-orange** [comment]
On ubuntu 20.04, for me the README still doesn't work or is broken again, and I didn't find anything above in the comments that works, can you help? Thanks

```bash
 $ curl -s "https://gitlab-ci-local-ppa.firecow.dk/pubkey.gpg" | sudo apt-key add -
OK
```
```bash
 $ echo "deb https://gitlab-ci-local-ppa.firecow.dk ./" | sudo tee /etc/apt/sources.list.d/gitlab-ci-local.list
deb https://gitlab-ci-local-ppa.firecow.dk ./
```
``` 
$ sudo apt-get update
Hit:1 http://us.archive.ubuntu.com/ubuntu focal InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease              
Hit:3 http://security.ubuntu.com/ubuntu focal-security InRelease    
Get:4 https://gitlab-ci-local-ppa.firecow.dk ./ InRelease [2,107 B] 
Hit:5 http://us.archive.ubuntu.com/ubuntu focal-updates InRelease
Hit:6 http://us.archive.ubuntu.com/ubuntu focal-backports InRelease
Err:4 https://gitlab-ci-local-ppa.firecow.dk ./ InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4A1691F3875016E
Reading package lists... Done
W: GPG error: https://gitlab-ci-local-ppa.firecow.dk ./ InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4A1691F3875016E
E: The repository 'https://gitlab-ci-local-ppa.firecow.dk ./ InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
